src/async_result.ts

```diff
@@ -0,0 +1,133 @@
+import Result, { Err, Ok } from "./result";
+import Monad from "./monad";
+
+type MaybePromise<T> = T | PromiseLike<T>;
+
+/**
+ * A hybrid of a `Result` and a `Promise`, `AsyncResult` includes
+ * the common convenience methods of working with a `Result` such
+ * as `map`, `bimap` and `flatMap`, while representing possibly-
+ * asynchronous state like a `Promise`.
+ *
+ * Aside from the convenience methods, `AsyncResult` also encodes
+ * type information about its error variant, unlike a native
+ * `Promise`, which may reject with `any` value.
+ */
+class AsyncResult<L, R> implements PromiseLike<Result<L, R>>, Monad<R> {
+  private promise: Promise<Result<L, R>>;
+
+  constructor(result: Result<L, R> | PromiseLike<Result<L, R>>) {
+    this.promise = Promise.resolve(result);
+  }
+
+  // Implement `PromiseLike<Result<L, R>>`
+  public then<TResult1 = Result<L, R>, TResult2 = never>(
+    onfulfilled?:
+      | ((value: Result<L, R>) => TResult1 | PromiseLike<TResult1>)
+      | null
+      | undefined,
```

Comment:
Do we need the optional types when we have `?`?

Reply:
Tbh I didn't look too closely at this signature after letting the tooling generate it for me. We could drop `| undefined` since that's redundant with `?`, but I think we're stuck with `| null` to be a valid implementer of the interface